### ISSUE TITLE
Why avoid gradient accumulation?

### ISSUE THREAD

**Issue by RonanKMcGovern** [google-research/tuning_playbook#69]
There is this quote:
```

**Gradient accumulation** simulates a larger batch size than the
--
252 | hardware can support and therefore does not provide any throughput
253 | benefits. It should generally be avoided in applied work.

```

For large GPUs and multi-GPU setups, I can see this making sense, as you can run batches of 32 and don't need accumulation.

Am I mistaken or missing something?

But, on smaller GPUs, grad accum can be important because it provides averaging in the virtual batches that stabilises the training.

**Comment by DimitrisMantas** [comment]
A lot of architecture have BN layers which don't work properly unless actually backprogated through, I think.

**Comment by RonanKMcGovern** [comment]
Interesting? What is BN? Bias?

When you say "a lot of", does that include Llama 2 and 3 type models?

Basically you're saying that accumulating the gradients isn't enough, some
important info is thrown away once you move to the next forward pass?

**Comment by DimitrisMantas** [comment]
Batch normalization. Essentially, BN blocks keep track of the running batch mean and standard deviation and use them to normalize their inputs.

These parameters are non-trainable and are updated with each minibatch they receive. However, because the total number of batches per epoch is not the same as that of backpropagations when using gradient accumulation, BN blocks compute "incorrect" statistics. This problem is further magnified by their other parameters still being updated according to accumulated batches. Basically, batches and their descriptive statistics become “unsynchronized”.

BN blocks are very popular in computer vision tasks, and unfortunately, I’m not too familiar with much else. However, I believe that transformer blocks use layer normalization blocks which do not depend on batch size.

**Comment by DimitrisMantas** [comment]
By the way, large batch sizes are just as "dangerous" as small ones due to potential overmoothing of the gradient landscape. It's kind of a "pick your poison" situation.

**Comment by RonanKMcGovern** [comment]
Thanks yeah agreed on the problems at big batches.

And yeah that makes sense re ga and batch norm. Llama 2 and 3 are layer
norm so should be fine but good to know for multi model models - I need to
check if clip has batch norm.
